java/ql/lib/semmle/code/java/dataflow/ListOfConstantsSanitizer.qll

```diff
@@ -0,0 +1,357 @@
+/**
+ * Provides classes for identifying comparisons against a list of compile-time
+ * constants and considering them as taint-sanitizers.
+ */
+
+import java
+private import semmle.code.java.controlflow.Guards
+private import semmle.code.java.dataflow.TaintTracking
+private import internal.DataFlowPrivate
+
+/**
+ * A comparison against a list of compile-time constants, sanitizing taint by
+ * restricting to a set of known values.
+ */
+class ListOfConstantsComparisonSanitizerGuard extends TaintTracking::DefaultTaintSanitizer {
+  ListOfConstantsComparisonSanitizerGuard() {
+    this = DataFlow::BarrierGuard<listOfConstantsComparisonSanitizerGuard/3>::getABarrierNode()
+  }
+}
+
+private predicate listOfConstantsComparisonSanitizerGuard(Guard g, Expr e, boolean outcome) {
+  exists(ListOfConstantsComparison locc |
+    g = locc and
+    e = locc.getExpr() and
+    outcome = locc.getOutcome()
+  )
+}
+
+private Expr skipCaseChanges(MethodCall mc) {
+  exists(Method changecase | mc.getMethod() = changecase |
+    changecase.hasName(["toUpperCase", "toLowerCase"]) and
+    changecase.getDeclaringType() instanceof TypeString
+  ) and
+  result = mc.getQualifier()
+}
+
+/** A comparison against a list of compile-time constants. */
+abstract class ListOfConstantsComparison extends Guard {
```

Comment:
Make private?

Reply:
My thinking in not making this all private was that someone could add some classes representing their favourite way of making a list of constants.